src/hotspot/share/memory/metaspaceShared.cpp

```diff
@@ -1612,34 +1613,50 @@ char* MetaspaceShared::reserve_address_space_for_archives(FileMapInfo* static_ma
       align_up(archive_space_size + gap_size + class_space_size,
                os::vm_allocation_granularity());
 
-  ReservedSpace total_rs;
-  if (base_address != NULL) {
-    // Reserve at the given archive base address, or not at all.
-    total_rs = ReservedSpace(total_range_size, archive_space_alignment,
-                             false /* bool large */, (char*) base_address);
+  assert(total_range_size > ccs_begin_offset, "must be");
+  if (use_windows_memory_mapping() && use_archive_base_addr) {
+    if (base_address != nullptr) {
+      address ccs_base = base_address + archive_space_size + gap_size;
+      archive_space_rs = ReservedSpace(archive_space_size, archive_space_alignment,
+                                       false /* large */, (char*)base_address);
+      class_space_rs   = ReservedSpace(class_space_size, class_space_alignment,
+                                       false /* large */, (char*)ccs_base);
+    }
+    if (!archive_space_rs.is_reserved() || !class_space_rs.is_reserved()) {
+      release_reserved_spaces(archive_space_rs, class_space_rs);
+      return NULL;
+    }
   } else {
```

Comment:
Could we have a comment explaining all this? I am sure I will have forgotten why we did this in two months :) Proposal:

```
On Windows, we cannot safely split a reserved area into two (see JDK-xxxx). Hence, 
 we optimistically reserve archive space and class space side-by-side. We only do this for
 use_archive_base_addr=true since for the use_archive_base_addr=false case caller will
 not split the combined space but read the archive data via sequential file IO.
```
Feel free to reformulate this.